go-controller/pkg/clustermanager/healthcheck/controller.go

```diff
@@ -0,0 +1,426 @@
+		Informer:       nodeInformer.Informer(),
+		Lister:         nodeInformer.Lister().List,
+		ObjNeedsUpdate: func(oldObj, newObj *corev1.Node) bool { return true },
+		Reconcile:      func(key string) error { return c.reconcile(key) },
+		Threadiness:    1,
+	}
+	controller := controllerutil.NewController[corev1.Node]("health check controller", config)
+
+	c.controller = controller
+	c.nodeInformer = nodeInformer
+	c.ctx = ctx
+
+	err := controllerutil.StartControllers(controller)
+	if err != nil {
+		return fmt.Errorf("failed to start health check controller: %w", err)
+	}
+
+	// probing loop, executes probing tasks from the queue, runs to
+	// completion
+	go func() {
+		c.run()
+
+		// stop & cleanup
+		controllerutil.StopControllers(controller)
+		c.cleanup()
+		klog.Infof("Stopped health check controller")
+	}()
+
+	klog.Infof("Started health check controller")
+	return nil
+}
+
+func (c *controller) run() {
+	for {
+		// pop the next probe task to execute
+		task := c.probeTaskQueue.Pop(c.ctx)
+		select {
+		case <-c.ctx.Done():
+			return
+		default:
+			// execute on its own goroutine
+			go c.probe(task)
+		}
+	}
+}
+
+func (c *controller) probe(task *probeTask) {
+	// check if task was cancelled before probing
+	nodeInfo, exists := c.isProbedWithNodeInfo(task.node)
+	if !exists || task.cancelled() {
+		task.client.Disconnect()
+		return
+	}
+
+	// probe
+	klog.V(5).Infof("Probing health state of node %s", task.node)
+	state := AVAILABLE
+	reachable := IsReachable(task.ctx, task.node, nodeInfo.ips, task.client, c.port, period)
+	if !reachable {
+		klog.Errorf("Failed health state probe for node %s", task.node)
+		state = UNREACHABLE
+	}
+
+	// update state & retask
+	updated := c.setNodeStateAndRetask(task, state)
+	if updated {
+		c.informConsumers(task.node)
+	}
+
+	// if no interest from consumers, reconcile to cancel the probing task
+	if !c.hasInterestedConsumers(task.node, nil) {
+		c.controller.Reconcile(task.node)
+	}
+}
+
+func (c *controller) reconcile(nodeName string) error {
+	start := time.Now()
+	klog.V(5).Infof("Reconciling health check for node %s", nodeName)
+	defer func() {
+		klog.V(5).Infof("Finished reconciling health check for node %s: %v", nodeName, time.Since(start))
+	}()
+
+	node, err := c.nodeInformer.Lister().Get(nodeName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if node == nil {
+		// node deleted, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: node deleted", nodeName)
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	names := []string{}
+	if !c.hasInterestedConsumers(nodeName, &names) {
+		// no interest, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: no interested consumers out of %q", nodeName, names)
+			// in case someone just registered
+			c.informConsumers(nodeName)
```

Comment:
if someone registerd cause they are interested, and you are shutting down the probe cause of no interest, how do you force this to reconcile again? Shouldn't we return an error here to force a retry?

Reply:
Not sure I understand. We are already in the reconciliation loop here. If I shutdown the probe because there is no interest, then what do I need to retry for? If someone gains interest again, then controller `MonitorHealthState` will be called which will trigger the reconciliation loop again.

Reply:
If there are no interested consumers, then why do you need to inform consumers? The comment says in case someone  just registered. I thought this means someone registers after you have already decided to remove the probe because you thought there were no interested consumers. So if I use your logic, then the there should be no reason to call informConsumers during this reconcile because it will happen when the reconciliation loop happens again.

Reply:
We inform any registered consumer of any health state change of any node:
```
	// HealthStateChanged informs consumers of node health state changes. A
	// consumer might be informed about a node it is not interested in and
	// should be able to ignore it in that case.
	HealthStateChanged(node string)
```

The interest just dictates whether a node should be probed or not.

So even if a node probing stops because there is no interest, it is the duty of the controller to inform any registered consumer that the state changed to UNKNOWN.

The comment might be misleading I guess so I will remove or rephrase.